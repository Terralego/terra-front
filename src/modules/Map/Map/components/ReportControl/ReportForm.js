import React, { useState, useCallback } from 'react';
import { Button, Label, RadioGroup, Radio, TextArea } from '@blueprintjs/core';

export const ReportForm = ({ url, coordinates, onSubmit, onCancel, translate: t }) => {
  const [comment, setComment] = useState('');
  const [reportType, setReportType] = useState('wrong_info');

  const onCommentChange = useCallback(e => setComment(e.currentTarget.value), []);
  const onReportTypeChange = useCallback(e => setReportType(e.currentTarget.value), []);

  const submitReport = useCallback(e => {
    e.preventDefault();
    onSubmit({ comment, coordinates, url, reportType });
  }, [onSubmit, comment, coordinates, url, reportType]);

  return (
    <form onSubmit={submitReport}>
      <input hidden id="report-geometry" value={coordinates} />
      <RadioGroup
        label={t('map.report_control.form.radiogroup.label')}
        selectedValue={reportType}
        onChange={onReportTypeChange}
      >
        <Radio
          label={t('map.report_control.form.radio.wrong_info.label')}
          value="wrong_info"
        />
        <Radio
          label={t('map.report_control.form.radio.missing_info.label')}
          value="missing_info"
        />
        <Radio
          label={t('map.report_control.form.radio.wrong_location.label')}
          value="wrong_location"
        />
        <Radio
          label={t('map.report_control.form.radio.doesnotexist.label')}
          value="does_not_exist"
        />
      </RadioGroup>
      <Label>
        {t('map.report_control.form.comment.label')}
        <span className="bp3-text-muted"> ({t('map.report_control.form.comment.optional')})</span>
        <TextArea value={comment} onChange={onCommentChange} />
      </Label>
      <Label>
        {t('map.report_control.form.url.label')}:
        <br />
        <a href={url}>{url}</a>
      </Label>

      <Button onClick={onCancel}>{t('map.report_control.form.cancel')}</Button>
      <Button type="submit" id="submitReport">{t('map.report_control.form.submit')}</Button>
    </form>
  );
};
export default ReportForm;
