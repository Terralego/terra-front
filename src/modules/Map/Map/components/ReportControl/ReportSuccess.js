import React from 'react';
import { Button } from '@blueprintjs/core';

const ReportSucess = ({ newReport, endReport, translate: t }) => (
  <div>
    <h1>{t('terralego.map.report_control.success.title')}</h1>
    <p>{t('terralego.map.report_control.success.message')}</p>
    <Button onClick={newReport}>{t('terralego.map.report_control.success.other_report')}</Button>
    <Button onClick={endReport}>{t('terralego.map.report_control.success.done')}</Button>
  </div>
);

export default ReportSucess;
