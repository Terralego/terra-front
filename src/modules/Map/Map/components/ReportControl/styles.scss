.report-card {
    position: absolute;
    top: 50%;
    left: 50%;
    /* bring your own prefixes */
    transform: translate(-50%, -50%);
  }
