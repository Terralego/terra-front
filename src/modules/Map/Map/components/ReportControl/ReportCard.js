import React, { useState, useCallback } from 'react';
import { Overlay, Card, Elevation } from '@blueprintjs/core';

import ReportForm from './ReportForm';
import ReportSuccess from './ReportSuccess';

export const ReportCard = ({
  isOpen,
  onSubmit,
  cancelReport,
  coordinates,
  reportUrl,
  newReport,
  endReport,
  translate: t,
}) => {
  const [submited, setSubmited] = useState(false);

  const onNewReport = useCallback(() => {
    setSubmited(false);
    newReport();
  }, [newReport]);

  const submitReport = useCallback(value => {
    setSubmited(true);
    onSubmit(value);
  }, [onSubmit]);


  return (
    <Overlay
      autoFocus
      lazy
      canOutsideClickClose
      canEscapeKeyClose
      isOpen={isOpen}
      onClose={cancelReport}
    >
      <Card
        interactive
        elevation={Elevation.FOUR}
        className="report-card"
      >
        {(submited && <ReportSuccess newReport={onNewReport} endReport={endReport} translate={t} />)
          || (
            <ReportForm
              onSubmit={submitReport}
              onCancel={cancelReport}
              coordinates={coordinates}
              url={reportUrl}
              translate={t}
            />
          )}
      </Card>
    </Overlay>
  );
};

export default ReportCard;
