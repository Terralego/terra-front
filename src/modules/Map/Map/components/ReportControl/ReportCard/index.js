import React, { useState } from 'react';
import { Overlay, Card, Elevation } from '@blueprintjs/core';

import ReportForm from '../ReportForm';
import ReportSuccess from '../ReportSuccess';

import './styles.scss';

export const ReportCard = ({
  isOpen,
  onSubmit,
  cancelReport,
  coordinates,
  reportUrl,
  newReport,
  endReport,
  translate: t,
}) => {
  const [submited, setSubmited] = useState(false);

  const title = submited ?
    'terralego.map.report_control.success.title' :
    'terralego.map.report_control.form.title';

  const onNewReport = () => {
    setSubmited(false);
    newReport();
  };

  const onEndReport = () => {
    setSubmited(false);
    endReport();
  };

  const submitReport = data => {
    setSubmited(true);
    onSubmit(data);
  };


  return (
    <Overlay
      autoFocus
      lazy
      canEscapeKeyClose
      isOpen={isOpen}
      onClose={cancelReport}
      className="overlay-container"
    >
      <Card
        interactive
        elevation={Elevation.FOUR}
        className="report-card"
      >
        <h1>{t(title)}</h1>
        {submited
          && <ReportSuccess newReport={onNewReport} endReport={onEndReport} translate={t} />}

        {!submited
        && (
        <ReportForm
          onSubmit={submitReport}
          onCancel={cancelReport}
          coordinates={coordinates}
          url={reportUrl}
          translate={t}
        />
        )}
      </Card>
    </Overlay>
  );
};

export default ReportCard;
