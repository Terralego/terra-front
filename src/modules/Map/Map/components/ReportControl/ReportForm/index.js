import React from 'react';
import { Button, FormGroup, InputGroup, Radio, TextArea } from '@blueprintjs/core';

import './styles.scss';

export const ReportForm = ({ url, coordinates, onSubmit, onCancel, translate: t }) => {
  const submitReport = e => {
    e.preventDefault();
    const data = {};
    const fd = new FormData(e.target);
    fd.forEach((value, key) => { data[key] = value; });
    onSubmit(data);
  };

  return (
    <form onSubmit={submitReport} onReset={onCancel}>
      <input hidden name="lng" value={coordinates.lng} />
      <input hidden name="lat" value={coordinates.lat} />
      <FormGroup
        label={t('terralego.map.report_control.form.radiogroup.label')}
      >
        <Radio
          label={t('terralego.map.report_control.form.radio.wrong_info.label')}
          value="wrong_info"
          name="reportType"
          defaultChecked
        />
        <Radio
          label={t('terralego.map.report_control.form.radio.missing_info.label')}
          value="missing_info"
          name="reportType"
        />
        <Radio
          label={t('terralego.map.report_control.form.radio.wrong_location.label')}
          value="wrong_location"
          name="reportType"
        />
        <Radio
          label={t('terralego.map.report_control.form.radio.doesnotexist.label')}
          value="does_not_exist"
          name="reportType"
        />
      </FormGroup>
      <FormGroup>
        {t('terralego.map.report_control.form.comment.label')}
        <span className="bp3-text-muted"> ({t('terralego.map.report_control.form.comment.optional')})</span>
        <TextArea large fill name="comment" />
      </FormGroup>
      <FormGroup className="report-url">
        {t('terralego.map.report_control.form.url.label')}:
        <InputGroup large readOnly name="url" defaultValue={url} />
      </FormGroup>

      <FormGroup className="report-button-group">
        <Button
          type="reset"
          className="report-button"
        >
          {t('terralego.map.report_control.form.cancel')}
        </Button>
        <Button
          type="submit"
          id="submitReport"
          className="report-button"
        >
          {t('terralego.map.report_control.form.submit')}
        </Button>
      </FormGroup>
    </form>
  );
};
export default ReportForm;
