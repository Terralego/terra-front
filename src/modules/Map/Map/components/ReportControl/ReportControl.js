import React from 'react';
import { Button, Icon } from '@blueprintjs/core';
import { Marker } from 'mapbox-gl';

import AbstractControl from '../../../helpers/AbstractMapControl';
import Tooltip from '../../../../../components/Tooltip';
import ReportCard from './ReportCard';


export default class ReportControl extends AbstractControl {
  static containerClassName = 'mapboxgl-ctrl mapboxgl-ctrl-group mapboxgl-ctrl-report';

  constructor (props) {
    super(props);
    this.state = {
      isReporting: false,
      coordinates: null,
      url: null,
    };
  }

  componentDidMount () {
    const {
      fromReport,
      map,
      reportCoords: { lat, lng },
    } = this.props;
    if (fromReport) {
      const coords = { lat: parseFloat(lat), lng: parseFloat(lng) };
      map.on('load', () => {
        this.reportMarker = new Marker()
          .setLngLat(coords)
          .addTo(map);
        map.flyTo({ center: coords });
      });
    }
  }

  componentWillUnmount () {
    if (this.reportMarker) {
      this.reportMarker.remove();
    }
  }

  toggleReport = e => {
    const { url, map } = this.props;
    const coordinates = e.lngLat;
    this.marker = new Marker().setLngLat(coordinates).addTo(this.props.map);
    this.setState({
      coordinates,
      isReporting: true,
      url,
    });

    // allowing just one click, prevent some conflict
    map.off('click', this.toggleReport);
  }

  onToggleReport = () => {
    const { map } = this.props;
    map.on('click', this.toggleReport);
  }

  onStopReport = () => {
    const { map } = this.props;
    map.off('click', this.toggleReport);
    this.marker.remove();
    this.setState({ isReporting: false, coordinates: null, url: null });
  }

  onNewReport = () => {
    this.onStopReport();
    this.onToggleReport();
  }

  render () {
    const { isReporting, coordinates = {} } = this.state;
    const { submitReport, translate: t, url } = this.props;

    return (
      <>
        <ReportCard
          isOpen={isReporting}
          cancelReport={this.onStopReport}
          endReport={this.onStopReport}
          newReport={this.onNewReport}
          onSubmit={submitReport}
          coordinates={coordinates}
          translate={t}
          reportUrl={coordinates && `${url}/${coordinates.lng}/${coordinates.lat}`}
        />
        <Tooltip
          content={t('terralego.map.report_control.content')}
        >
          <Button
            onClick={this.onToggleReport}
          >
            <Icon icon="error" />
          </Button>
        </Tooltip>
      </>
    );
  }
}
