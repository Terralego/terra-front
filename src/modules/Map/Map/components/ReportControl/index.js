import  ReportControl from './ReportControl';

export default ReportControl;
