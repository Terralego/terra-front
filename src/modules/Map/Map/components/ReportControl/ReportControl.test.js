import React from 'react';
import renderer from 'react-test-renderer';

import ReportControl from './ReportControl';

jest.mock('react-dom', () => ({
  createPortal: () => <div>Blueprint3.Overlay</div>, // fix usePortal error of Blueprint overlay
  render: jest.fn(),
}));

jest.mock('mapbox-gl', () => ({
  Marker: () => ({}),
}));

it('should render correctly', () => {
  const tree = renderer.create(
    <ReportControl
      translate={jest.fn()}
      reportCoords={{ lat: '44,4', lng: '4.55' }}
    />,
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it('should render correctly when reporting', () => {
  const component = renderer.create(
    <ReportControl
      translate={jest.fn()}
      reportCoords={{ lat: '44,4', lng: '4.55' }}
    />,
  );
  component.getInstance().setState({ isReporting: true });
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('should add eventlistener on map when toggling report', () => {
  const mockedMap = { on: jest.fn() };
  const component = renderer.create(
    <ReportControl
      translate={jest.fn()}
      map={mockedMap}
      reportCoords={{ lat: '44,4', lng: '4.55' }}
    />,
  );
  const instance = component.getInstance();
  instance.displayToaster = jest.fn();
  instance.onToggleReport();
  expect(mockedMap.on).toHaveBeenCalled();
});

it('should remove eventlistener and marker on map when stopping report', () => {
  const mockedMap = { off: jest.fn() };
  const mockedMarker = { remove: jest.fn() };
  const component = renderer.create(
    <ReportControl
      translate={jest.fn()}
      map={mockedMap}
      reportCoords={{ lat: '44,4', lng: '4.55' }}
    />,
  );
  // mock this.marker.remove()
  // marker is removed when reporting is stopped
  component.getInstance().marker = mockedMarker;
  component.getInstance().onStopReport();
  expect(mockedMap.off).toHaveBeenCalled();
  expect(mockedMarker.remove).toHaveBeenCalled();
});

it('should stop report and toggle a new one when asking new report', () => {
  const mockedStopReport = jest.fn();
  const mockedToggleReport = jest.fn();
  const component = renderer.create(
    <ReportControl
      translate={jest.fn()}
      map={{ on: jest.fn(), off: jest.fn() }}
      reportCoords={{ lat: '44,4', lng: '4.55' }}
    />,
  );
  const instance = component.getInstance();
  instance.marker = { remove: jest.fn() };
  instance.onStopReport = mockedStopReport;
  instance.onToggleReport = mockedToggleReport;
  instance.onNewReport();
  expect(mockedStopReport).toHaveBeenCalled();
  expect(mockedToggleReport).toHaveBeenCalled();
});
