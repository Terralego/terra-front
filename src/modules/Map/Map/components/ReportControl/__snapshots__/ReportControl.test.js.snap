// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<span
  className="bp3-popover-wrapper"
>
  <span
    className="bp3-popover-target"
    onBlur={[Function]}
    onFocus={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <button
      className="bp3-button"
      onClick={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="bp3-button-text"
      >
        <span
          className="bp3-icon bp3-icon-error"
          icon="error"
        >
          <svg
            data-icon="error"
            height={16}
            viewBox="0 0 16 16"
            width={16}
          >
            <desc>
              error
            </desc>
            <path
              d="M7.99-.01c-4.42 0-8 3.58-8 8s3.58 8 8 8 8-3.58 8-8-3.58-8-8-8zm1 13h-2v-2h2v2zm0-3h-2v-7h2v7z"
              fillRule="evenodd"
            />
          </svg>
        </span>
      </span>
    </button>
  </span>
</span>
`;

exports[`should render correctly when reporting 1`] = `
Array [
  <div>
    Blueprint3.Overlay
  </div>,
  <span
    className="bp3-popover-wrapper"
  >
    <span
      className="bp3-popover-target"
      onBlur={[Function]}
      onFocus={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <button
        className="bp3-button"
        onClick={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="bp3-button-text"
        >
          <span
            className="bp3-icon bp3-icon-error"
            icon="error"
          >
            <svg
              data-icon="error"
              height={16}
              viewBox="0 0 16 16"
              width={16}
            >
              <desc>
                error
              </desc>
              <path
                d="M7.99-.01c-4.42 0-8 3.58-8 8s3.58 8 8 8 8-3.58 8-8-3.58-8-8-8zm1 13h-2v-2h2v2zm0-3h-2v-7h2v7z"
                fillRule="evenodd"
              />
            </svg>
          </span>
        </span>
      </button>
    </span>
  </span>,
]
`;
